langchain_tools_demo/tools.py

```diff
@@ -0,0 +1,256 @@
+
+
+@tool(
+    "Get Airport",
+    args_schema=AirportIdInput,
+)
+def get_airport(id: int):
+    """
+    Use this tool to get info for a specific airport.
+    Do NOT guess an airport id.
+    Takes an id and returns info on the airport.
+    """
+    response = get_request(
+        f"{BASE_URL}/airports",
+        {"id": id},
+    )
+    if response.status_code != 200:
+        return f"Error trying to find airport: {response.text}"
+
+    return response.json()
+
+
+class FlightIdInput(BaseModel):
+    id: int = Field(description="Unique identifier")
+
+
+@tool(
+    "Get Flight",
+    args_schema=FlightIdInput,
+)
+def get_flight(id: int):
+    """
+    Use this tool to get info for a specific flight.
+    Takes an id and returns info on the flight.
+    Do NOT use this tool if you have a flight number.
+    Do NOT guess an airline or flight number.
+    """
+    response = get_request(
+        f"{BASE_URL}/flights",
+        {"flight_id": id},
+    )
+    if response.status_code != 200:
+        return f"Error trying to find flight: {response.text}"
+
+    return response.json()
+
+
+class FlightNumberInput(BaseModel):
+    airline: str = Field(description="Airline unique 2 letter identifier")
+    flight_number: str = Field(description="1 to 4 digit number")
+
+
+@tool(
+    "Get Flights by Number",
+    args_schema=FlightNumberInput,
+)
+def search_flights_by_number(airline: str, flight_number: str):
+    """
+    Use this tool to get info for a specific flight. Do NOT use this tool with a flight id.
+    Takes an airline and flight number and returns info on the flight.
+    Do NOT guess an airline or flight number.
+    A flight number is a code for an airline service consisting of two-character
+    airline designator and a 1 to 4 digit number ex. OO123, DL 1234, BA 405, AS 3452.
+    If the tool returns more than one option choose the date closes to today.
+    """
+    response = get_request(
+        f"{BASE_URL}/flights/search",
+        {"airline": airline, "flight_number": flight_number},
+    )
+    if response.status_code != 200:
+        return f"Error trying to find flight: {response.text}"
+
+    return response.json()
+
+
+class ListFlights(BaseModel):
+    departure_airport: Optional[str] = Field(
+        description="Departure airport 3-letter code",
+    )
+    arrival_airport: Optional[str] = Field(description="Arrival airport 3-letter code")
+    date: Optional[str] = Field(description="Date of flight departure")
+
+
+@tool("List Flights", args_schema=ListFlights)
+def list_flights(departure_airport: str, arrival_airport: str, date: str):
+    """
+    Use this tool to list all flights matching search criteria.
+    Takes an arrival airport and departure airport and returns all flights.
+    The agent can decide to return the results directly to the user.
+    Input of this tool must be in JSON format and include all three inputs - arrival_airport, departure_airport, and date.
```

Comment:
The search flight endpoint supports search with only `arrival_airport` or only `departure_airport` and doesn't require both. Do we want to disallow that for the agent? The user may want to search for "all flights leaving SFO today".

Reply:
This aims at getting the agent to include all three inputs even if they are null. We would have to subclass BaseTool in order to fix the validation if the input, even though optional, isn't included.